src/3d/mesh/qgsmesh3dmaterial_p.cpp

```diff
@@ -220,3 +311,56 @@ void QgsMesh3dMaterial::configure()
   mTechnique->addParameter( new Qt3DRender::QParameter( "isScalarMagnitude", ( mMagnitudeType == QgsMesh3dMaterial::ScalarDataSet ) ) );
 }
 
+void QgsMesh3dMaterial::configureArrows( QgsMeshLayer *layer )
+{
+  if ( !layer )
+    return;
+
+  QgsMeshDatasetIndex datasetIndex = layer->rendererSettings().activeVectorDataset();
+
+  mTechnique->addParameter( new Qt3DRender::QParameter( "arrowsSpacing", float( mSymbol.arrowsSpacing() ) ) ) ;
+  QColor arrowsColor = layer->rendererSettings().vectorSettings( datasetIndex.group() ).color();
+  mTechnique->addParameter( new Qt3DRender::QParameter( "arrowsColor", QVector4D( arrowsColor.redF(), arrowsColor.greenF(), arrowsColor.blueF(), 1.0f ) ) ) ;
+
+  QgsMeshDatasetGroupMetadata meta = layer->dataProvider()->datasetGroupMetadata( datasetIndex );
+  mTechnique->addParameter( new Qt3DRender::QParameter( "maxVectorLength", float( meta.maximum() ) ) ) ;
+
+  QVector<QgsVector> vectors;
+  QSize gridSize;
+  QgsPointXY minCorner;
+  Qt3DRender::QParameter *arrowsEnabledParameter = new Qt3DRender::QParameter( "arrowsEnabled", false );
+  if ( mMagnitudeType != MagnitudeType::ScalarDataSet || !mSymbol.arrowsEnabled() )
+    arrowsEnabledParameter->setValue( false );
+  else
+  {
+    arrowsEnabledParameter->setValue( true );
+    gridSize = QSize( mSymbol.maximumTextureSize(), mSymbol.maximumTextureSize() );
+    vectors = QgsMeshLayerUtils::griddedVectorValues(
+                layer,
+                datasetIndex,
+                mSymbol.arrowsSpacing(),
+                mSymbol.arrowsSpacing(),
+                gridSize,
+                minCorner );
```

Comment:
Hmm I would prefer to split griddedVectorValues() into two parts: the first one would find out the final grid size, the second one would just prepare the values.

And as Peter noted, it would be good to reuse the existing code for gridded vector values if possible, so that we do not have two slightly different versions.